### PR TITLE
feat(replays): add recordings payload ItemType

### PR DESCRIPTION
**Pull request by JoshFerge**
- Adds an ItemType for replay recordings 
- Logic to set the ItemType based on the name of the file in multitype logic
- Same flow as attachments, but sent to a different topic
- An ItemType was chosen as opposed to another attachment type as we'll likely begin to diverge from attachments logic and this gives us more flexibility long term.

See the spec here for more background information: 
https://www.notion.so/sentry/Session-Replay-V1-alpha-Ingest-Backend-ae068d1e1d514221b6c3ea2233f360f4

----

By submitting this pull request, I confirm that Sentry can use, modify, copy, and redistribute this contribution, under Sentry's choice of terms.

### CODE REVIEW

File: relay-server/src/utils/multipart.rs [hunk 1]
```diff
@@ -218,7 +220,12 @@ fn consume_item(
         let file_name = content_disposition.as_ref().and_then(|d| d.get_filename());
 
         if let Some(file_name) = file_name {
-            let mut item = Item::new(ItemType::Attachment);
+            let item_type = match file_name {
+                REPLAY_RECORDING_FILENAME => ItemType::ReplayRecording,
```
Comment (JoshFerge) [inline relay-server/src/utils/multipart.rs]: there should only ever be 1 replay attachment per envelope, not sure if we should take that into account here / other parts where it can accept N.

We could also make replay recording uploads a separate endpoint altogether.

File: tests/integration/fixtures/processing.py [hunk 2]
```diff
@@ -334,3 +340,23 @@ def get_individual_attachment(self):
...
+
+
+class ReplayRecordingsConsumer(EventsConsumer):
+    def get_replay_chunk(self):
+        message = self.poll()
+        assert message is not None
+        assert message.error() is None
+
+        v = msgpack.unpackb(message.value(), raw=False, use_list=False)
+        assert v["type"] == "attachment_chunk", v["type"]
```
Comment (JoshFerge) [inline tests/integration/fixtures/processing.py]: would probably want to have this type be `replay_chunk` for cleanliness sake, as well as attachment_type be something else, but for now i don't think these matter.

File: relay-server/src/envelope.rs [hunk 3]
```diff
@@ -142,6 +144,7 @@ impl fmt::Display for ItemType {
             Self::MetricBuckets => write!(f, "metric_buckets"),
             Self::ClientReport => write!(f, "client_report"),
             Self::Profile => write!(f, "profile"),
+            Self::ReplayPayload => write!(f, "replay_payload"),
```
Comment (jan-auer) [inline relay-server/src/envelope.rs]: Please also add an entry for `replay_payload` to `FromStr` below and add a roundtrip test at the bottom of the file.

Comment (jan-auer) [inline relay-server/src/envelope.rs]: In envelopes, we use `payload` to refer to the contents of an item. Using "payload" here is slightly misleading, could we find a more descriptive name to differentiate the replay data from its event?

File: relay-server/src/envelope.rs [hunk 4]
```diff
@@ -647,6 +651,7 @@ impl Item {
             ItemType::Metrics => false,
             ItemType::MetricBuckets => false,
             ItemType::ClientReport => false,
+            ItemType::ReplayPayload => false,
```
Comment (jan-auer) [inline relay-server/src/envelope.rs]: Note: This is correct as long as https://github.com/getsentry/relay/pull/1239 doesn't make this an event type, but rather a completely separate payload.

File: relay-server/src/actors/store.rs [hunk 5]
```diff
@@ -714,11 +728,14 @@ impl Handler<StoreEnvelope> for StoreForwarder {
             KafkaTopic::Attachments
         } else if event_item.map(|x| x.ty()) == Some(&ItemType::Transaction) {
             KafkaTopic::Transactions
+        } else if envelope.get_item_by(is_replay_payload).is_some() {
+            KafkaTopic::ReplayPayloads
```
Comment (jan-auer) [inline relay-server/src/actors/store.rs]: Please revert this. These topic assignments are only meant for event payloads and their dependent items (such as user feedback). You cannot and should not produce events to the replay topic either.

File: relay-server/src/actors/store.rs [hunk 6]
```diff
@@ -728,6 +745,7 @@ impl Handler<StoreEnvelope> for StoreForwarder {
                         event_id.ok_or(StoreError::NoEventId)?,
                         scoping.project_id,
                         item,
+                        topic,
```
Comment (jan-auer) [inline relay-server/src/actors/store.rs]: Same here, please revert this change and continue to write attachment chunks to the attachments topic.

File: relay-server/src/actors/store.rs [hunk 7]
```diff
@@ -762,6 +780,16 @@ impl Handler<StoreEnvelope> for StoreForwarder {
...
                 )?,
+                ItemType::ReplayPayload => {
+                    debug_assert!(topic == KafkaTopic::ReplayPayloads);
+                    let replay_payload = self.produce_attachment_chunks(
+                        event_id.ok_or(StoreError::NoEventId)?,
+                        scoping.project_id,
+                        item,
+                        topic,
+                    )?;
+                    replay_payloads.push(replay_payload);
```
Comment (jan-auer) [inline relay-server/src/actors/store.rs]: Instead of collecting the replay payloads into a vector and producing them at the end, produce them directly as you're iterating. 

With the current logic, you would silently drop replay payloads if there happens to be an event in the same envelope (which should be supported from SDK side imho).

File: relay-server/src/actors/store.rs [hunk 8]
```diff
@@ -797,6 +825,21 @@ impl Handler<StoreEnvelope> for StoreForwarder {
                     event_type = "attachment"
                 );
             }
+        } else if !replay_payloads.is_empty() {
+            relay_log::trace!("Sending individual replay_payloads of envelope to kafka");
+            for attachment in replay_payloads {
+                let replay_payload_message = KafkaMessage::ReplayPayload(AttachmentKafkaMessage {
```
Comment (jan-auer) [inline relay-server/src/actors/store.rs]: Rather than reusing `AttachmentKafkaMessage`, please create a dedicated `ReplayPayload` message with all the attributes you need. This also allows you to name the replay ID attribute something like `session_id` or `replay_id`.

File: relay-server/src/actors/store.rs [hunk 9]
```diff
@@ -762,6 +780,16 @@ impl Handler<StoreEnvelope> for StoreForwarder {
                     start_time,
                     item,
                 )?,
+                ItemType::ReplayPayload => {
+                    debug_assert!(topic == KafkaTopic::ReplayPayloads);
+                    let replay_payload = self.produce_attachment_chunks(
```
Comment (jan-auer) [inline relay-server/src/actors/store.rs]: What is the expected (max) payload size of one replay payload? The default chunk size we produce to Kafka is 1MB, so if your clients are already chunking the payloads you can also directly produce here. 

Of course, in this case, we'd have to add some validation for this at an earlier point. Let me know if that sounds reasonable and I'll point to the places.

The way you've currently implemented this suggests that the client buffers the payload of the _entire_ replay and sends it in one go.

File: relay-server/src/actors/store.rs [hunk 10]
```diff
@@ -690,6 +700,10 @@ fn is_slow_item(item: &Item) -> bool {
     item.ty() == &ItemType::Attachment || item.ty() == &ItemType::UserReport
 }
 
+fn is_replay_payload(item: &Item) -> bool {
```
Comment (jan-auer) [inline relay-server/src/actors/store.rs]: _nit_: Inline

File: relay-server/src/actors/store.rs [hunk 11]
```diff
@@ -648,6 +657,7 @@ impl KafkaMessage {
             Self::Session(_message) => Uuid::nil(), // Explicit random partitioning for sessions
             Self::Metric(_message) => Uuid::nil(),  // TODO(ja): Determine a partitioning key
             Self::Profile(_message) => Uuid::nil(),
+            Self::ReplayPayload(_message) => Uuid::nil(),
```
Comment (jan-auer) [inline relay-server/src/actors/store.rs]: Using NIL here means you're not guaranteed to write all replay messages to the same Kafka partition, which means no guaranteed order of delivery. 

From our recent conversations, this should be fine. However, the implementation below produces the `AttachmentChunks` and `Attachment` messages, which require order. See the comment below and if you keep attachment messages use a partitioning key here.

File: relay-server/src/utils/multipart.rs [hunk 12]
```diff
@@ -218,7 +220,12 @@ fn consume_item(
         let file_name = content_disposition.as_ref().and_then(|d| d.get_filename());
 
         if let Some(file_name) = file_name {
-            let mut item = Item::new(ItemType::Attachment);
+            let item_type = match file_name {
+                REPLAY_RECORDING_FILENAME => ItemType::ReplayPayload,
+                _ => ItemType::Attachment,
+            };
+            let mut item = Item::new(item_type);
```
Comment (jan-auer) [inline relay-server/src/utils/multipart.rs]: All new SDKs are expected to send Envelopes rather than multipart/form-data. We should not add more capabilities to the multipart endpoints. Hence, please use the Envelope facilities in the JavaScript SDK to send an Envelope with the appropriate ItemType directly.

The changes in this file can then be reverted.

File: relay-config/src/config.rs [hunk 13]
```diff
@@ -830,6 +835,7 @@ impl Default for TopicAssignments {
             sessions: "ingest-sessions".to_owned().into(),
             metrics: "ingest-metrics".to_owned().into(),
             profiles: "profiles".to_owned().into(),
+            replay_payloads: "ingest-replay-payloads".to_owned().into(),
```
Comment (jan-auer) [inline relay-config/src/config.rs]: Before deploying, please ensure the topic is provisioned in the broker.

File: relay-server/src/utils/sizes.rs [hunk 14]
```diff
@@ -27,7 +27,7 @@ pub fn check_envelope_size_limits(config: &Config, envelope: &Envelope) -> bool
             | ItemType::Security
             | ItemType::RawSecurity
             | ItemType::FormData => event_size += item.len(),
-            ItemType::Attachment | ItemType::UnrealReport => {
+            ItemType::Attachment | ItemType::UnrealReport | ItemType::ReplayPayload => {
```
Comment (jan-auer) [inline relay-server/src/utils/sizes.rs]: Since replays will become their own thing (including for rate limiting), consider having separate limits for them. This relates to one of the comments in the store actor, since you may want to have smaller limits per replay item (such as 1MB). The limit for attachments is at 100MB right now.

File: tests/integration/test_replay_payloads.py [hunk 15]
```diff
@@ -0,0 +1,60 @@
+import pytest
+import time
+import uuid
+
+from requests.exceptions import HTTPError
+
+
+def test_payload(
+    mini_sentry, relay_with_processing, replay_payloads_consumer, outcomes_consumer
```
Comment (jan-auer) [inline tests/integration/test_replay_payloads.py]: Please also add one test for a relay without processing mode. See here: https://github.com/getsentry/relay/blob/604fe06d517a2109a6b6ed85935c9641fb3848ff/tests/integration/test_session.py#L9

**Review by jan-auer — CHANGES_REQUESTED** [review]
Thank you, initial review attached. We had two general thoughts when reading through this change:

- Right now, this PR does not add any rate limiting capabilities and also does not track data categories. This is perfectly fine for the initial implementation. Once you want to check for rate limits or track outcomes, you'll also have to add a new `DataCategory`.
- Since this is experimental, it would be good to add a feature flag that allows us to disable and drop replays. For an example of how to do this, see `process_profiling`:

https://github.com/getsentry/relay/blob/604fe06d517a2109a6b6ed85935c9641fb3848ff/relay-server/src/actors/envelopes.rs#L975-L976

File: relay-server/src/utils/multipart.rs (hunk 1 above)
Comment (jan-auer) [reply relay-server/src/utils/multipart.rs]: For this you could return `true` from `is_duplicate`. What's the reason behind this constraint? It could be beneficial for clients with offline caching to send several replay payloads in a single envelope.

File: relay-server/src/envelope.rs (hunk 3 above)
Comment (JoshFerge) [reply relay-server/src/envelope.rs]: yeah we discussed this internally as a team as well, going to rename to `ReplayRecording`.

File: relay-server/src/actors/store.rs (hunk 9 above)
Comment (JoshFerge) [reply relay-server/src/actors/store.rs]: while the client is already sending payloads in chunks, unfortunately, an _individual chunk_ can still be several megabytes, in the case of a checkout where we need to send a lot of DOM information, e.g., so i think we'll still need to produce chunk messages here. 

We could engineer the frontend to do its own splitting of requests within chunks it needs to send, but i'd rather push this complexity to here as opposed to being on the sdk (plus the extra network requests it would require)

File: relay-server/src/actors/store.rs [hunk 16]
```diff
@@ -439,6 +446,54 @@ impl StoreForwarder {
...
+        let id = Uuid::new_v4().to_string();
+
+        let mut chunk_index = 0;
+        let mut offset = 0;
+        let payload = item.payload();
+        let size = item.len();
+
+        // This skips chunks for empty replay recordings. The consumer does not require chunks for
+        // empty replay recordings. `chunks` will be `0` in this case.
+        while offset < size {
```
Comment (JoshFerge) [inline relay-server/src/actors/store.rs:463]: these bits are copied from the attachment chunks produce function.

File: relay-server/src/actors/store.rs [hunk 17]
```diff
@@ -439,6 +446,54 @@ impl StoreForwarder {
...
+            offset += chunk_size;
+            chunk_index += 1;
+        }
+
+        // The chunk_index is incremented after every loop iteration. After we exit the loop, it
+        // is one larger than the last chunk, so it is equal to the number of chunks.
+
+        Ok(ChunkedAttachment {
+            id,
+            name: REPLAY_RECORDINGS_ATTACHMENT_NAME.to_owned(),
```
Comment (JoshFerge) [inline relay-server/src/actors/store.rs]: i don't forsee a use for the `name` field for us, but didn't seem worth defining our own type just to remove it.

**Comment by JoshFerge** [comment]
alright ready for re-review. added comments where i had questions / thoughts. only other thing is if using `event_id` as the field in the envelope headers is okay, or if we should add replay_id as a top level header field.

**Comment by JoshFerge** [comment]
hmm not sure why the integration test is failing, passes on my machine.

File: CHANGELOG.md [hunk 18]
```diff
@@ -43,6 +43,7 @@
 - Refactor aggregation error, recover from errors more gracefully. ([#1240](https://github.com/getsentry/relay/pull/1240))
 - Remove/reject nul-bytes from metric strings. ([#1235](https://github.com/getsentry/relay/pull/1235))
 - Remove the unused "internal" data category. ([#1245](https://github.com/getsentry/relay/pull/1245))
+- Add ReplayRecording ItemType. ([#1236](https://github.com/getsentry/relay/pull/1236))
```
Comment (jan-auer) [inline CHANGELOG.md]: Note: Please move the changelog entry to the "Unreleased" section just before merging this.

File: relay-server/src/actors/envelopes.rs [hunk 19]
```diff
@@ -1006,6 +1006,19 @@ impl EnvelopeProcessor {
...
 
+    fn process_replay_recordings(&self, state: &mut ProcessEnvelopeState) {
+        let replays_enabled = state.project_state.has_feature(Feature::Replays);
+        state.envelope.retain_items(|item| match item.ty() {
+            ItemType::ReplayRecording => {
+                if !replays_enabled {
+                    return false;
+                }
+                true
+            }
```
Comment (jan-auer) [inline relay-server/src/actors/envelopes.rs]: _nit_: Can be shortened to:

```suggestion
            ItemType::ReplayRecording => replays_enabled,
```

File: relay-server/src/actors/store.rs (hunk 17 above)
Comment (jan-auer) [reply relay-server/src/actors/store.rs]: Keeping some constant name here is a good solution to distinguish the message contents in other places 👍 

If all you need here are `id`, `chunks`, and `size`, then you can also add those fields directly to your `ReplayRecordingKafkaMessage`, though.

File: relay-server/src/actors/store.rs (hunk 16 above)
Comment (jan-auer) [reply relay-server/src/actors/store.rs:463]: Thanks for calling this out. For now this is totally fine, once we have a third use case, we can still generalize the chunking implementation.

File: relay-server/src/actors/store.rs [hunk 20]
```diff
@@ -558,6 +613,32 @@ struct AttachmentKafkaMessage {
...
+struct ReplayRecordingChunkKafkaMessage {
+    /// Chunk payload of the replay recording.
+    payload: Bytes,
+    /// The replay id.
+    replay_id: EventId,
+    /// The project id for the current replay.
+    project_id: ProjectId,
+    /// The recording ID within the replay.
+    ///
+    /// The triple `(project_id, replay_id, id)` identifies a replay recording chunk uniquely.
```
Comment (jan-auer) [inline relay-server/src/actors/store.rs]: Is this actually true? The `id` alone is a unique identifier for the replay recording, but it's shared by all chunks. The tuple `(id, chunk_index)` is the unique identifier for a single chunk, plus it determines ordering.

**Review by jan-auer — APPROVED** [review]
Thank you for the thorough work and applying all feedback. This looks good and is ready to merge from my end. See minor stylistic comments below.

File: relay-server/src/actors/store.rs (hunk 20 above)
Comment (JoshFerge) [reply relay-server/src/actors/store.rs]: you are correct, fixing comment.

File: relay-server/src/actors/store.rs (hunk 17 above)
Comment (JoshFerge) [reply relay-server/src/actors/store.rs]: just went ahead and defined a separate type with just these 3 fields 👍🏼
